tokio-util/src/sync/cancellation_token.rs

```diff
@@ -241,6 +241,22 @@ impl CancellationToken {
     pub fn drop_guard(self) -> DropGuard {
         DropGuard { inner: Some(self) }
     }
+
+    /// Runs a future to completion and returns its result unless the `CompletionToken` is
+    /// cancelled. In that case the function returns `None` and the future gets dropped.
+    ///
+    /// # Cancel safety
+    ///
+    /// This method is not cancel safe since the input future gets dropped on cancel.
+    pub async fn run_until_cancelled<F>(&self, fut: F) -> Option<F::Output>
+    where
+        F: std::future::Future,
+    {
+        tokio::select! {
+            _ = self.cancelled() => None,
+            fut_result = fut => Some(fut_result),
+        }
```

Comment:
This should at least use a biased select.
```suggestion
        tokio::select! {
            biased;
            _ = self.cancelled() => None,
            fut_result = fut => Some(fut_result),
        }
```
But we can also avoid the `select!` macro entirely (and in turn, avoid the macros feature) by implementing the `Future` trait manually. Is that something you would be interested in? It is not absolutely necessary to get this PR merged.

Reply:
Of course I would be interested in that. Sounds reasonable to do that to keep the dependencies low. I will give it a try and than we can still discuss if this PR should get merged or not.

You mean, we should return a custom type that implements `Future` and return that (just like `WaitForCancellationFuture` I would call it `RunUntilCancelledFuture`).

Reply:
Let's not make it publicly visible. You can define a private struct and then just `.await` it inside the function. I recommend that you start with `tokio::time::timeout` as a starting point:

https://github.com/tokio-rs/tokio/blob/833ee027d0ec44d88765157ec98b6809f0070169/tokio/src/time/timeout.rs#L190-L228

However, you don't need any of the coop budget stuff, so you can get rid of that.

Reply:
Thanks for the guidance. I have submitted some changes that should incorporate your thoughts on this.